pkg/proxy/proxy.go

```diff
@@ -430,36 +429,35 @@ func (p *Proxy) ReinstallRoutingRules() error {
 			return err
 		}
 
-		if err := removeFromEgressProxyRoutesIPv6(); err != nil {
-			return err
-		}
-		if requireFromProxyRoutes() {
+		if fromIngressProxy || fromEgressProxy {
 			ipv6, err := getCiliumNetIPv6()
 			if err != nil {
 				return err
 			}
-			if err := installFromProxyRoutesIPv6(ipv6, defaults.HostDevice); err != nil {
+			if err := installFromProxyRoutesIPv6(ipv6, defaults.HostDevice, fromIngressProxy, fromEgressProxy); err != nil {
 				return err
 			}
 		} else {
-			if err := removeFromIngressProxyRoutesIPv6(); err != nil {
+			if err := removeFromProxyRoutesIPv6(); err != nil {
 				return err
 			}
 		}
 	} else {
 		if err := removeToProxyRoutesIPv6(); err != nil {
 			return err
 		}
-		if err := removeFromIngressProxyRoutesIPv6(); err != nil {
+		if err := removeFromProxyRoutesIPv6(); err != nil {
 			return err
 		}
 	}
 
 	return nil
 }
 
-func requireFromProxyRoutes() bool {
-	return (option.Config.EnableEnvoyConfig || option.Config.EnableIPSec) && !option.Config.TunnelingEnabled()
+func requireFromProxyRoutes() (fromIngressProxy, fromEgressProxy bool) {
+	fromIngressProxy = (option.Config.EnableEnvoyConfig || option.Config.EnableIPSec) && !option.Config.TunnelingEnabled()
+	fromEgressProxy = option.Config.EnableIPSec && !option.Config.TunnelingEnabled()
+	return
```

Comment:
I'm still a bit bugged by this. IIUC, removing the IPsec condition here causes the CI to fail. So it breaks something when IPsec is disabled, but we should trust IPsec fixes it? It would be the first time IPsec fixes something :sweat_smile:

Reply:
:laughing: Actually, the complete storyline is:
1. https://github.com/cilium/cilium/pull/29530 installs `0xa00 lookup 2005` when `ipsec=on`
2. https://github.com/cilium/cilium/pull/32367 extends the condition of installing above routing to `envoy=on || ipsec=on`
3. this pr wants to install a new routing `0xb00 lookup 2005`. If we keep using the condition `envoy=on || ipsec=on`, the new routing will also be installed when envoy=on && ipsec=off, but some errors in ci-e2e (non-ipsec) were seen.  Therefore I decide to limit the new routing to ipsec only, here we are, two different conditions.

Reply:
I was mostly wondering if we can make this code become:
```go
fromEgressProxy = !option.Config.TunnelingEnabled()
```
Or even just:
```go
fromEgressProxy = true
```
:sweat_smile: 

But I guess then we'd pay the price of one more redirect per packet so it's probably not worth it...

Reply:
I think we're definitely due for a wider re-design of the via-proxy traffic paths, to eliminate all the current special cases. Maybe introduce IPsec logic in `to-netdev` instead, and not rely on `bpf_host` so much?

But that's most likely not backport-able, and it's been tricky enough just to get *this* part working :grimacing:.